server/src/main/java/org/elasticsearch/search/aggregations/support/BuiltinValuesSourceType.java

```diff
@@ -0,0 +1,214 @@
+            }
+
+            ValuesSource.Numeric dataSource = new ValuesSource.Numeric.FieldData((IndexNumericFieldData)fieldContext.indexFieldData());
+            if (script != null) {
+                // What's the difference between ValuesSource.Numeric.Script and ValuesSource.Numeric.WithScript?
+                dataSource = new ValuesSource.Numeric.WithScript(dataSource, script);
+            }
+            return dataSource;
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            Number missing = docValueFormat.parseDouble(rawMissing.toString(), false, now);
+            return MissingValues.replaceMissing((ValuesSource.Numeric) valuesSource, missing);
+        }
+    },
+    BYTES {
+        @Override
+        public ValuesSource getEmpty() {
+            return ValuesSource.Bytes.WithOrdinals.EMPTY;
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            return new ValuesSource.Bytes.Script(script);
+        }
+
+        @Override
+        public ValuesSource getField(FieldContext fieldContext, AggregationScript.LeafFactory script) {
+            final IndexFieldData<?> indexFieldData = fieldContext.indexFieldData();
+            ValuesSource dataSource;
+            if (indexFieldData instanceof IndexOrdinalsFieldData) {
+                dataSource = new ValuesSource.Bytes.WithOrdinals.FieldData((IndexOrdinalsFieldData) indexFieldData);
+            } else {
+                dataSource = new ValuesSource.Bytes.FieldData(indexFieldData);
+            }
+            if (script != null) {
+                // Again, what's the difference between WithScript and Script?
+                dataSource = new ValuesSource.Bytes.WithScript(dataSource, script);
+            }
+            return dataSource;
+        }
+
+        @Override
+        public ValuesSource replaceMissing(ValuesSource valuesSource, Object rawMissing, DocValueFormat docValueFormat, LongSupplier now) {
+            final BytesRef missing = docValueFormat.parseBytesRef(rawMissing.toString());
+            if (valuesSource instanceof ValuesSource.Bytes.WithOrdinals) {
+                return MissingValues.replaceMissing((ValuesSource.Bytes.WithOrdinals) valuesSource, missing);
+            } else {
+                return MissingValues.replaceMissing((ValuesSource.Bytes) valuesSource, missing);
+            }
+        }
+    },
+    GEOPOINT {
+        @Override
+        public ValuesSource getEmpty() {
+            return ValuesSource.GeoPoint.EMPTY;
+        }
+
+        @Override
+        public ValuesSource getScript(AggregationScript.LeafFactory script, ValueType scriptValueType) {
+            throw new AggregationExecutionException("value source of type [" + this.value() + "] is not supported by scripts");
```

Comment:
I wonder if we should make this exception the same as the other `getFoo()` methods, whatever we decide to keep it consistent?

Reply:
See comment above.  Changing exception types in this PR is too many moving parts IMHO.